### PR TITLE
add/edit stellar address in farm

### PR DESCRIPTION
**Pull request by AlaaElattar**
- https://github.com/threefoldtech/threefold_connect/issues/730
- User can add or edit stellar address of farm
- also user can choose address from saved wallets
- address always being validated, won't be saved unless its correct.

### Tested Scenarios:
- Tried to save wrong address but failed as button is disabled.
- Tried to save right address and worked.
- Chose address from wallets and worked.

### CODE REVIEW

**Comment by AlaaElattar** [comment]
Uploading 2024-11-27 13.13.03.mp4…

**Review by zaelgohary — CHANGES_REQUESTED** [review]
I could edit a stellar address with an invalid value. I think we should validate the address before adding/editing.

**Review by zaelgohary — COMMENTED** [review]
After clicking edit, there is no option to cancel the operation. To cancel editing, I would need to click save.

image

**Comment by zaelgohary** [comment]
Clicking edit and navigating to the homepage without saving leaves the farming page stuck in the editing state.

**Comment by AlaaElattar** [comment]
> After clicking edit, there is no option to cancel the operation. To cancel editing, I would need to click save.
> 
> image

- Done, added cancel button that appears while editing.

**Comment by AlaaElattar** [comment]
> Clicking edit and navigating to the homepage without saving leaves the farming page stuck in the editing state.

tried to reproduce it but didn't happen

File: app/lib/widgets/farm_item.dart [hunk 1]
```diff
@@ -29,9 +47,85 @@ class _FarmItemWidgetState extends State<FarmItemWidget> {
...
+      setState(() {
+        edit = false;
+        isSaving = false;
+      });
+    }
+  }
+
+  void validateStellarAddress(String address) {
+    setState(() {
+      addressError =
```
Comment (AhmedHanafy725) [inline app/lib/widgets/farm_item.dart:112]: can you also check if the address has TFT trustline?

File: app/lib/widgets/farm_item.dart [hunk 2]
```diff
@@ -48,24 +142,97 @@ class _FarmItemWidgetState extends State<FarmItemWidget> {
...
+                  labelText: 'Stellar Payout Address',
+                  suffixIcon: edit
+                      ? IconButton(
+                          onPressed: () {
+                            Navigator.of(context).push(MaterialPageRoute(
+                              builder: (context) => ContractsScreen(
+                                  chainType: chainType,
+                                  currentWalletAddress:
+                                      widget.farm.walletAddress,
+                                  wallets: widget.wallets,
```
Comment (AhmedHanafy725) [inline app/lib/widgets/farm_item.dart]: can you exclude the wallets with `-1` balance?

File: app/lib/widgets/farm_item.dart [hunk 3]
```diff
@@ -48,24 +142,97 @@ class _FarmItemWidgetState extends State<FarmItemWidget> {
...
+              decoration: InputDecoration(
+                  errorText: addressError,
+                  labelText: 'Stellar Payout Address',
+                  suffixIcon: edit
+                      ? IconButton(
+                          onPressed: () {
+                            Navigator.of(context).push(MaterialPageRoute(
+                              builder: (context) => ContractsScreen(
+                                  chainType: chainType,
+                                  currentWalletAddress:
```
Comment (AhmedHanafy725) [inline app/lib/widgets/farm_item.dart]: after a successful update, the `widget.farm.walletAddress` is still the same as the old one, it should be updated to not show the current wallet in the contacts list

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]

**Review by zaelgohary — APPROVED** [review]

File: app/lib/widgets/farm_item.dart [hunk 4]
```diff
@@ -29,9 +49,103 @@ class _FarmItemWidgetState extends State<FarmItemWidget> {
...
+    } finally {
+      setState(() {
+        edit = false;
+        isSaving = false;
+      });
+    }
+  }
+
+  void validateStellarAddress(String address) async {
+    setState(() {
```
Comment (AhmedHanafy725) [inline app/lib/widgets/farm_item.dart:111]: please trim the address before start validating

File: app/lib/widgets/farm_item.dart [hunk 5]
```diff
@@ -29,9 +49,103 @@ class _FarmItemWidgetState extends State<FarmItemWidget> {
...
+        }
+      } catch (e) {
+        setState(() {
+          addressError = 'Error fetching account balance';
+        });
+      }
+    }
+  }
+
+  void _selectToAddress(String address) {
```
Comment (AhmedHanafy725) [inline app/lib/widgets/farm_item.dart]:
```suggestion
  void _selectAddress(String address) {

```

File: app/lib/widgets/farm_item.dart [hunk 6]
```diff
@@ -29,9 +49,103 @@ class _FarmItemWidgetState extends State<FarmItemWidget> {
...
+    setState(() {
+      walletAddressController.text = address;
+      validateStellarAddress(address);
```
Comment (AhmedHanafy725) [inline app/lib/widgets/farm_item.dart]: I don't think we should validate the address here

**Review by AhmedHanafy725 — APPROVED** [review]

**Pull request merged**